### ISSUE TITLE
Trouble with iptables rules on Synology

### ISSUE THREAD

**Issue by frysztak** [erikespinoza/v4raider#2]
first of all, thank you so much for creating this repo. it was incredibly helpful in bootstrapping my self-hosting setup.

after setting everything up, it turned out that traffic directed at VPS does get routed through WireGuard tunnel, but anything coming out of the tunnel doesn't end up at eth0. following [this](https://www.digitalocean.com/community/tutorials/how-to-forward-ports-through-a-linux-gateway-with-iptables) tutorial, I replaced iptables' rules with those:
```
PostUp   = iptables -A FORWARD -i eth0 -o %i -p tcp --syn -m multiport --dports 80,443 -m conntrack --ctstate NEW -j ACCEPT
PostDown = iptables -D FORWARD -i eth0 -o %i -p tcp --syn -m multiport --dports 80,443 -m conntrack --ctstate NEW -j ACCEPT

PostUp   = iptables -A FORWARD -i eth0 -o %i -m conntrack --ctstate ESTABLISHED,RELATED -j ACCEPT
PostDown = iptables -D FORWARD -i eth0 -o %i -m conntrack --ctstate ESTABLISHED,RELATED -j ACCEPT

PostUp   = iptables -A FORWARD -i %i -o eth0 -m conntrack --ctstate ESTABLISHED,RELATED -j ACCEPT
PostDown = iptables -D FORWARD -i %i -o eth0 -m conntrack --ctstate ESTABLISHED,RELATED -j ACCEPT

PostUp   = iptables -t nat -A PREROUTING -i eth0 -p tcp -m multiport --dports 80,443 -j DNAT --to-destination 10.1.0.2
PostDown = iptables -t nat -D PREROUTING -i eth0 -p tcp -m multiport --dports 80,443 -j DNAT --to-destination 10.1.0.2

PostUp   = iptables -t nat -A POSTROUTING -o %i -p tcp -m multiport --dports 80,443 -d 10.1.0.2 -j SNAT --to-source 10.1.0.1
PostDown = iptables -t nat -D POSTROUTING -o %i -p tcp -m multiport --dports 80,443 -d 10.1.0.2 -j SNAT --to-source 10.1.0.1
```

and it works great! I believe the last SNAT rule was the culprit. is it intentionally missing from `rules.v4` or is there something wrong with my setup?

**Comment by erikespinoza** [comment]
Let me start by saying that adding an SNAT sounds like a good thing. No objections to that. If you send a pull request, happy to include it or I'll get to it later in the week.

That said, I am unsure I recognize what you mean. I am literally using these rules on a few hosts and things are working correctly. I would really like to understand what went wrong. Mind giving me a test case or more info? For example a cmd line example with curl and where you are curl'ing from?

Thanks for the feedback and I'm glad this helped you out.

**Comment by frysztak** [comment]
thanks for replying so quickly. here's what happens when go `curl -I VPSIP` using original rules generated by your script (I'm running those commands on my VPS):
```
$ sudo tcpdump -n -v -i eth0 port 80
tcpdump: listening on eth0, link-type EN10MB (Ethernet), capture size 262144 bytes
09:44:00.164585 IP (tos 0x0, ttl 52, id 10900, offset 0, flags [DF], proto TCP (6), length 60)
    <WANIP>.35288 > <VPSIP>.80: Flags [S], cksum 0x9efa (correct), seq 3072849193, win 64240, options [mss 1460,sackOK,TS val 2571637662 ecr 0,nop,wscale 7], length 0
09:44:01.236489 IP (tos 0x0, ttl 52, id 10901, offset 0, flags [DF], proto TCP (6), length 60)
    <WANIP>.35288 > <VPSIP>.80: Flags [S], cksum 0x9aca (correct), seq 3072849193, win 64240, options [mss 1460,sackOK,TS val 2571638734 ecr 0,nop,wscale 7], length 0
09:44:03.316317 IP (tos 0x0, ttl 52, id 10902, offset 0, flags [DF], proto TCP (6), length 60)
    <WANIP>.35288 > <VPSIP>.80: Flags [S], cksum 0x92aa (correct), seq 3072849193, win 64240, options [mss 1460,sackOK,TS val 2571640814 ecr 0,nop,wscale 7], length 0
09:44:07.396151 IP (tos 0x0, ttl 52, id 10903, offset 0, flags [DF], proto TCP (6), length 60)
    <WANIP>.35288 > <VPSIP>.80: Flags [S], cksum 0x82ba (correct), seq 3072849193, win 64240, options [mss 1460,sackOK,TS val 2571644894 ecr 0,nop,wscale 7], length 0
09:44:15.475660 IP (tos 0x0, ttl 52, id 10904, offset 0, flags [DF], proto TCP (6), length 60)
    <WANIP>.35288 > <VPSIP>.80: Flags [S], cksum 0x632a (correct), seq 3072849193, win 64240, options [mss 1460,sackOK,TS val 2571652974 ecr 0,nop,wscale 7], length 0
```

and

```
$ sudo tcpdump -n -v -i wg0 port 80
tcpdump: listening on wg0, link-type RAW (Raw IP), capture size 262144 bytes
09:44:00.164648 IP (tos 0x0, ttl 51, id 10900, offset 0, flags [DF], proto TCP (6), length 60)
    <WANIP>.35288 > 10.1.0.2.80: Flags [S], cksum 0x6352 (correct), seq 3072849193, win 64240, options [mss 1460,sackOK,TS val 2571637662 ecr 0,nop,wscale 7], length 0
09:44:01.236552 IP (tos 0x0, ttl 51, id 10901, offset 0, flags [DF], proto TCP (6), length 60)
    <WANIP>.35288 > 10.1.0.2.80: Flags [S], cksum 0x5f22 (correct), seq 3072849193, win 64240, options [mss 1460,sackOK,TS val 2571638734 ecr 0,nop,wscale 7], length 0
09:44:03.316364 IP (tos 0x0, ttl 51, id 10902, offset 0, flags [DF], proto TCP (6), length 60)
    <WANIP>.35288 > 10.1.0.2.80: Flags [S], cksum 0x5702 (correct), seq 3072849193, win 64240, options [mss 1460,sackOK,TS val 2571640814 ecr 0,nop,wscale 7], length 0
09:44:07.396206 IP (tos 0x0, ttl 51, id 10903, offset 0, flags [DF], proto TCP (6), length 60)
    <WANIP>.35288 > 10.1.0.2.80: Flags [S], cksum 0x4712 (correct), seq 3072849193, win 64240, options [mss 1460,sackOK,TS val 2571644894 ecr 0,nop,wscale 7], length 0
09:44:15.475720 IP (tos 0x0, ttl 51, id 10904, offset 0, flags [DF], proto TCP (6), length 60)
    <WANIP>.35288 > 10.1.0.2.80: Flags [S], cksum 0x2782 (correct), seq 3072849193, win 64240, options [mss 1460,sackOK,TS val 2571652974 ecr 0,nop,wscale 7], length 0
```
`<WANIP>` is my PC's IP.
My suspicion here is that SYN packet gets into WG tunnel (because `AllowedIPs = 0.0.0.0/0` is set on client) but response doesn't get back. Here's what happens when I use my custom rules:
```
$ sudo tcpdump -n -v -i eth0 port 80
tcpdump: listening on eth0, link-type EN10MB (Ethernet), capture size 262144 bytes
09:40:10.886752 IP (tos 0x0, ttl 52, id 36895, offset 0, flags [DF], proto TCP (6), length 60)
    <WANIP>.46572 > <VPSIP>.80: Flags [S], cksum 0x3dca (correct), seq 870133051, win 64240, options [mss 1460,sackOK,TS val 2571408375 ecr 0,nop,wscale 7], length 0
09:40:10.913863 IP (tos 0x0, ttl 63, id 0, offset 0, flags [DF], proto TCP (6), length 60)
    <VPSIP>.80 > <WANIP>.46572: Flags [S.], cksum 0x5d01 (correct), seq 2521601542, ack 870133052, win 27360, options [mss 1380,sackOK,TS val 260715835 ecr 2571408375,nop,wscale 7], length 0
09:40:10.940433 IP (tos 0x0, ttl 52, id 36896, offset 0, flags [DF], proto TCP (6), length 52)
    <WANIP>.46572 > <VPSIP>.80: Flags [.], cksum 0xf431 (correct), ack 1, win 502, options [nop,nop,TS val 2571408429 ecr 260715835], length 0
09:40:10.940434 IP (tos 0x0, ttl 52, id 36897, offset 0, flags [DF], proto TCP (6), length 131)
    <WANIP>.46572 > <VPSIP>.80: Flags [P.], cksum 0x4c6c (correct), seq 1:80, ack 1, win 502, options [nop,nop,TS val 2571408429 ecr 260715835], length 79: HTTP, length: 79
        HEAD / HTTP/1.1
        Host: <VPSIP>
        User-Agent: curl/7.68.0
        Accept: */*

09:40:10.967100 IP (tos 0x0, ttl 63, id 41846, offset 0, flags [DF], proto TCP (6), length 52)
    <VPSIP>.80 > <WANIP>.46572: Flags [.], cksum 0xf4cc (correct), ack 80, win 214, options [nop,nop,TS val 260715889 ecr 2571408429], length 0
09:40:10.967655 IP (tos 0x0, ttl 63, id 41847, offset 0, flags [DF], proto TCP (6), length 220)
    <VPSIP>.80 > <WANIP>.46572: Flags [P.], cksum 0xb777 (correct), seq 1:169, ack 80, win 214, options [nop,nop,TS val 260715889 ecr 2571408429], length 168: HTTP, length: 168
        HTTP/1.1 308 Permanent Redirect
        Location: https://<VPSIP>/
        Date: Sat, 02 Jan 2021 08:40:10 GMT
        Content-Length: 18
        Content-Type: text/plain; charset=utf-8

09:40:10.994272 IP (tos 0x0, ttl 52, id 36898, offset 0, flags [DF], proto TCP (6), length 52)
    <WANIP>.46572 > <VPSIP>.80: Flags [.], cksum 0xf2cf (correct), ack 169, win 501, options [nop,nop,TS val 2571408483 ecr 260715889], length 0
09:40:10.994523 IP (tos 0x0, ttl 52, id 36899, offset 0, flags [DF], proto TCP (6), length 52)
    <WANIP>.46572 > <VPSIP>.80: Flags [F.], cksum 0xf2ce (correct), seq 80, ack 169, win 501, options [nop,nop,TS val 2571408483 ecr 260715889], length 0
09:40:11.021322 IP (tos 0x0, ttl 63, id 41848, offset 0, flags [DF], proto TCP (6), length 52)
    <VPSIP>.80 > <WANIP>.46572: Flags [F.], cksum 0xf3b6 (correct), seq 169, ack 81, win 214, options [nop,nop,TS val 260715943 ecr 2571408483], length 0
09:40:11.047722 IP (tos 0x0, ttl 52, id 36900, offset 0, flags [DF], proto TCP (6), length 52)
    <WANIP>.46572 > <VPSIP>.80: Flags [.], cksum 0xf262 (correct), ack 170, win 501, options [nop,nop,TS val 2571408536 ecr 260715943], length 0
```

and

```
$ sudo tcpdump -n -v -i wg0 port 80
tcpdump: listening on wg0, link-type RAW (Raw IP), capture size 262144 bytes
09:40:10.886857 IP (tos 0x0, ttl 51, id 36895, offset 0, flags [DF], proto TCP (6), length 60)
    10.1.0.1.46572 > 10.1.0.2.80: Flags [S], cksum 0xf479 (correct), seq 870133051, win 64240, options [mss 1460,sackOK,TS val 2571408375 ecr 0,nop,wscale 7], length 0
09:40:10.913816 IP (tos 0x0, ttl 64, id 0, offset 0, flags [DF], proto TCP (6), length 60)
    10.1.0.2.80 > 10.1.0.1.46572: Flags [S.], cksum 0x13b1 (correct), seq 2521601542, ack 870133052, win 27360, options [mss 1380,sackOK,TS val 260715835 ecr 2571408375,nop,wscale 7], length 0
09:40:10.940523 IP (tos 0x0, ttl 51, id 36896, offset 0, flags [DF], proto TCP (6), length 52)
    10.1.0.1.46572 > 10.1.0.2.80: Flags [.], cksum 0xaae1 (correct), ack 1, win 502, options [nop,nop,TS val 2571408429 ecr 260715835], length 0
09:40:10.940572 IP (tos 0x0, ttl 51, id 36897, offset 0, flags [DF], proto TCP (6), length 131)
    10.1.0.1.46572 > 10.1.0.2.80: Flags [P.], cksum 0x031c (correct), seq 1:80, ack 1, win 502, options [nop,nop,TS val 2571408429 ecr 260715835], length 79: HTTP, length: 79
        HEAD / HTTP/1.1
        Host: <VPSIP>
        User-Agent: curl/7.68.0
        Accept: */*

09:40:10.967057 IP (tos 0x0, ttl 64, id 41846, offset 0, flags [DF], proto TCP (6), length 52)
    10.1.0.2.80 > 10.1.0.1.46572: Flags [.], cksum 0xab7c (correct), ack 80, win 214, options [nop,nop,TS val 260715889 ecr 2571408429], length 0
09:40:10.967640 IP (tos 0x0, ttl 64, id 41847, offset 0, flags [DF], proto TCP (6), length 220)
    10.1.0.2.80 > 10.1.0.1.46572: Flags [P.], cksum 0x6e27 (correct), seq 1:169, ack 80, win 214, options [nop,nop,TS val 260715889 ecr 2571408429], length 168: HTTP, length: 168
        HTTP/1.1 308 Permanent Redirect
        Location: https://<VPSIP>/
        Date: Sat, 02 Jan 2021 08:40:10 GMT
        Content-Length: 18
        Content-Type: text/plain; charset=utf-8

09:40:10.994321 IP (tos 0x0, ttl 51, id 36898, offset 0, flags [DF], proto TCP (6), length 52)
    10.1.0.1.46572 > 10.1.0.2.80: Flags [.], cksum 0xa97f (correct), ack 169, win 501, options [nop,nop,TS val 2571408483 ecr 260715889], length 0
09:40:10.994536 IP (tos 0x0, ttl 51, id 36899, offset 0, flags [DF], proto TCP (6), length 52)
    10.1.0.1.46572 > 10.1.0.2.80: Flags [F.], cksum 0xa97e (correct), seq 80, ack 169, win 501, options [nop,nop,TS val 2571408483 ecr 260715889], length 0
09:40:11.021281 IP (tos 0x0, ttl 64, id 41848, offset 0, flags [DF], proto TCP (6), length 52)
    10.1.0.2.80 > 10.1.0.1.46572: Flags [F.], cksum 0xaa66 (correct), seq 169, ack 81, win 214, options [nop,nop,TS val 260715943 ecr 2571408483], length 0
09:40:11.047812 IP (tos 0x0, ttl 51, id 36900, offset 0, flags [DF], proto TCP (6), length 52)
    10.1.0.1.46572 > 10.1.0.2.80: Flags [.], cksum 0xa912 (correct), ack 170, win 501, options [nop,nop,TS val 2571408536 ecr 260715943], length 0
```

as you can see in `wg0` capture, both source and destination addresses got replaced by WG internal network IPs.

one more small thing: I needed to add 
```
dns:
      - 8.8.8.8
```
to wireguard service. otherwise Traefik wouldn't resolve my domain, and couldn't generate SSL certs. altough maybe this is a quirk of [synology-wireguard](https://github.com/runfalk/synology-wireguard#faqknown-issues)

**Comment by erikespinoza** [comment]
So this is meant to be a starting point. Sounds like it helped you, so I am happy about that. Setting dns is part of the included docker-compose.yml, but it sounds like you are using a setup that is somewhat different than what I had targeted.

I'm going to add the SNAT because I think its good for the definition to go both ways, but I suspect the problems were likely caused  by the differences in your setup.

Perhaps you can help me by sending me more details on the Synology specific divergence. That way we can help people who have your setup find answers more easily.

Thanks for engaging. Always happy to include details that help others.

**Comment by frysztak** [comment]
yes, it helped me immensely, so thank you again :).

about my Synology setup, I kept getting `iptables-restore v1.6.1: iptables-restore: unable to initialize table 'raw'` so I added
```
Table = 51820
PostUp = ip rule add to 10.1.0.0/24 lookup 51820
PostDown = ip rule del to 10.1.0.0/24 lookup 51820
```

to client wg0.conf (as described [here](https://github.com/runfalk/synology-wireguard/issues/36#issuecomment-677959599)). that and `dns` line are the only changes I made.
